opentelemetry-exporter-geneva/geneva-uploader/src/config_service/client.rs

```diff
@@ -0,0 +1,536 @@
+                        (&cached_data.auth_info, &cached_data.token_endpoint)
+                    {
+                        return Ok((ingestion.clone(), moniker.clone(), endpoint.clone()));
+                    }
+                }
+            }
+        }
+
+        // Cache miss or expired token, fetch fresh data
+        // Perform actual fetch before acquiring write lock to minimize lock contention
+        let (fresh_token_info, fresh_moniker_info) = self.fetch_ingestion_info().await?;
+        let token_expiry = self.parse_token_expiry(&fresh_token_info.auth_token_expiry_time);
+        let token_endpoint =
+            extract_endpoint_from_token(&fresh_token_info.auth_token).map_err(|e| {
+                GenevaConfigClientError::Certificate(format!(
+                    "Failed to extract endpoint from token: {}",
+                    e
+                ))
+            })?;
+
+        // Now update the cache with exclusive write access
+        {
+            let mut cached_data = self.cached_data.write().map_err(|_| {
+                GenevaConfigClientError::InternalError("RwLock poisoned".to_string())
+            })?;
+
+            // Double-check in case another thread updated while we were fetching
+            if let Some(existing) = cached_data.token_expiry {
+                if let Some(new) = token_expiry {
+                    if existing >= new {
+                        // Existing token is newer, use it
+                        if let (Some(info), Some(endpoint)) =
+                            (&cached_data.auth_info, &cached_data.token_endpoint)
+                        {
+                            return Ok((info.0.clone(), info.1.clone(), endpoint.clone()));
+                        }
+                    }
+                }
+            }
+
+            // Update cache with fresh data
+            cached_data.auth_info = Some((fresh_token_info, fresh_moniker_info));
+            cached_data.token_endpoint = Some(token_endpoint);
+            cached_data.token_expiry = token_expiry;
+
+            if let (Some(info), Some(endpoint)) =
+                (&cached_data.auth_info, &cached_data.token_endpoint)
+            {
+                Ok((info.0.clone(), info.1.clone(), endpoint.clone()))
+            } else {
+                Err(GenevaConfigClientError::InternalError(
+                    "Cache was not properly updated".to_string(),
+                ))
+            }
+        }
+    }
+
+    /// Internal method that actually fetches data from Geneva Config Service
+    async fn fetch_ingestion_info(&self) -> Result<(IngestionGatewayInfo, MonikerInfo)> {
+        let agent_identity = "GenevaUploader"; // TODO make this configurable
+        let agent_version = "0.1"; // TODO make this configurable
+        let identity = format!(
+            "Tenant=Default/Role=GcsClient/RoleInstance={}",
+            agent_identity
+        );
+
+        let encoded_identity = general_purpose::STANDARD.encode(&identity);
```

Comment:
This also looks like something that doesn't need to be computed for each call.

Reply:
Good catch - have precomputed them within GenevaConfigClient::new()